Linux wayland support

MAlba124 opened the pull request:
This implements basic wayland screencapture.

#### TODO

- [x] Select sources
- [x] Negotiate frame format
- [x] Receive video frames
- [x] Bitmap
- [x] Send the video frames to the user
- [x] Detect stalling when format negotiation fails
- [ ] GPU

MAlba124 committed:
init wayland

MAlba124 committed:
linux-wayland: fix test name typo

OutOfTheVoid commented:
This is looking awesome so far.

I think we may want to build an abstraction around whether window selection can happen application side vs window manager side, since the current interface isn't a 1:1 match with how it's done on xdg-desktop-portal, but that's still TBD.

I'm planning to also do a basic XCB based implementation which is more in line with the current API, but likely with lower performance. To that end, can you make the platform folder name represent the capturing backend as well? Something like `linux-xdg-desktop-portal` or `linux-wayland`.

Backend selection is probably another important abstraction we want to make for any case where there are multiple supported backends. For now, that's just a future plan for Linux.

MAlba124 committed:
linux-wayland: rename implementation directory

MAlba124 committed:
linux-wayland: send frame to callback

MAlba124 committed:
linux-wayland: handle disconnects and errors in stream

Callback for state_changed event so that stream errors and
unconnected events can be handled and notify consumer(s)

MAlba124 committed:
linux-wayland: handle unsupported content filter

+ treat virtual sources as display

MAlba124 committed:
linux-wayland: composit metadata cursor onto frame

MAlba124 committed:
linux-wayland: feature: bitmap

MAlba124 committed:
linux-wayland: bitmap: treat BGRx as BGRA

MAlba124 committed:
linux-wayland: disable cursor_as_metadata

MAlba124 commented:
I have removed the option to render the cursor as metadata as there seems to be a lot of edge cases where it won't work, e.g. when I tested display capture + cursor as meta on [plasma](https://kde.org/en/plasma-desktop/) the cursor was already rendered on the frame when it was received. This whole functionality can be omitted by having a show_cursor flag in `CaptureContentFilter`. This is preferred as it greatly reduces the implementation complexity.

There's also a problem with the red and blue channels should be swapped for some pixel formats on some hardware. This seems to be a known problem and there's probably a way to identify these cases. I would have to look into how OBS-studio does it.

MAlba124 commented:
Hey! Is this project sunset?
